### ISSUE TITLE
[FEATURE] Multiple py-envs in a single html file

### ISSUE THREAD

**Issue by nixjdm** [pyscript/pyscript#149]
Assuming it's possible, having multiple, different, simultaneously available py-envs could be very useful. As an easy example, consider two interactive demonstrations of different major versions of the same library, illustrating a new feature or breaking change. Or different versions of Python showing performance improvements with runnable tests.

There could be a one-to-many relationship between `<py-env>` tags and other py-script tags that use them. The one-to-one case would still be the most common, so existing html files could work as-is. For the others, we could allow an optional id to connect the tags. E.g. `<py-env id="foo">` could be used by name with both `<py-script py-env="foo">` and `<py-repl py-env="foo">` being present in the same html file, along with any additional groups like `id="bar"`.

**Comment by nixjdm** [comment]
Extending this idea a bit, if 1) Python in pyscript can interact with the DOM, and 2) multiple simultaneous runtimes can run just fine, then it seems like you could use multiple py-envs/py-scripts to write code that uses otherwise incompatible things, communicating over the DOM as a bus. Perhaps there's better ways to do that, but if that works that's an interesting door to open. It could in-effect allow API driven, environment and dependency independent "microservices."

Maybe my assumptions here are wrong though. I'm not very familiar with this WASM stuff yet.

**Comment by antocuni** [comment]
Thank you for the idea!
I might agree that it might be useful in some cases, but probably very hard to implement.
It's not even a pyscript problem, is a more general python problem. How would you implement such a thing in python? I imagine two different possibilities:
1. you somehow try to load two different versions of the same library in the same python process. 
2. you spawn multiple python processes, each with its own env, but then you cannot easily pass objects between the two.

Solution (1) looks doomed from the start to be, because python is not really designed for this.
With solution (2) you would have two completely separate instances of pyscript, which cannot (easily) communicate each other. You can probably achieve a similar result by having a single HTML with two independent `<iframe>`s.

I'm marking this as feature request; /cc @fpliger to see what are his thoughts and whether it should be a "wontfix" or not,.

**Comment by nixjdm** [comment]
What I'm thinking of is two separate processes, but not direct inter-process communication, rather, an indirect communication as with an external bus or information store. If it's possible to run multiple runtimes that can see the same DOM (i.e. in one html file), it should be possible to pass information back and forth through the DOM, just as two separate processes could over a file system, or something like Redis. I can imagine one processes setting information, and another polling for changes in a loop. To my understanding, Pyodide allows a Python WASM runtime to read and write from the DOM, so then that DOM IO part is basically taken care of. It's easy enough to set a variable in the DOM, and easy enough to read one in a loop.

Whether you can communicate between multiple runtimes then, to me, just hinges on whether you can actually have multiple runtimes at all. Perhaps this is more in Pyodide's court.

Assuming you could though, it would be valuable for PyScript to expose control over them. Even not considering deliberately passing information between them, if each runtime was attached to a specific py-env, then people could ship / incorporate py-scripts unedited, with no concern for resolving conflicts between them, like numpy or Python versions. It would make hacking things together from others' FOSS work much easier.

**Comment by antocuni** [comment]
Ok, I get the idea now.
Note that the two processes don't necessarily need to communicate through the DOM; there might be other IPC facilities available in WASM and/or emscripten, although I'm not really familiar with it.

**Comment by nixjdm** [comment]
Yes, that's right. Actual IPC would certainly have some benefits, but I don't know that that exists (or again, if you can even have multiple simultaneous WASM runtimes atm). There could be any kind of ways to communicate down the road. The DOM is just an easy to reach for early stage bus.

**Comment by fpliger** [comment]
So... if iirc, pyodide only support one runtime at the moment There is work being done to actually support more than one instance but it's not ready yet. (@rth  please correct me here if I'm mistaken...)

That said, this is an interesting idea though, especially considering that, in the future, we'd like to support more than just Pyodide and Python.

Not sure if we could take any action right now though

**Comment by rth** [comment]
> if iirc, pyodide only support one runtime at the moment There is work being done to actually support more than one instance but it's not ready yet.

So that feature was merged https://github.com/pyodide/pyodide/pull/2391 but it's indeed not released yet. An issue with using it  for multiple envs though is that, as far as I know, it would increase by up to 2x the load time. Since all packages would need to be re-installed (though wheel files should be taken from the browser cache) and nothing is currently shared between instances.

> What I'm thinking of is two separate processes, but not direct inter-process communication, rather, an indirect communication as with an external bus or information store.

That sounds more like a web-worker use case. You can start as many web-workers with separate Pyodide instances as necessary already now (though again load time scales with their number). The DOM interaction is a bit more difficult but https://github.com/hoodmane/synclink and https://github.com/alexmojaki/comsync aim to make interaction with workers easier.

**Comment by fpliger** [comment]
> So that feature was merged https://github.com/pyodide/pyodide/pull/2391 but it's indeed not released yet. An issue with using it for multiple envs though is that, as far as I know, it would increase by up to 2x the load time. Since all packages would need to be re-installed (though wheel files should be taken from the browser cache) and nothing is currently shared between instances. 

Ooooh... that's a bummer. I was definitely hoping that the loading time addition would have been only from new packages since the packages shared by environments would have run already.

**Comment by nixjdm** [comment]
If I wanted "microservices" out of pyodide, than web workers does sound like a pretty good fit. Reading https://pyodide.org/en/stable/usage/webworker.html#caveats, everything listed there sounds in fact desirable, minus the single threaded limitation.

Though, I also can still pretty easily see someone wanting to use multiple pyscripts that do interact with the browser UI, and still may need separate envs, or at least it would be nicer for the user to not worry about that and just give them separate envs. That ability to not worry about env conflicts is very valuable, imo. In that case it seems like we'd have to wait on https://github.com/pyodide/pyodide/pull/2391 being released.

@rth Thanks for pointing out web workers to me. It seems like there's a case for wanting multiple py-envs (for that sake of leveraging https://github.com/pyodide/pyodide/pull/2391), as well as good wrapper around web workers to easily define their envs, too.

**Comment by rth** [comment]
> I was definitely hoping that the loading time addition would have been only from new packages since the packages shared by environments would have run already.

So there could be different ways to approach this,
 1. creating a new Pyodide instance is similar in spirit to creating to launching separate VMs. One would get a separate Emscripten virtual filesystem, a separate Python but it's indeed more expensive to setup. We need to benchmark how much is the overhead exactly (and it likely depends how good is the browser at caching).
 2. taking the virtualenv analogy, it should be possible to keep a single Emscripten virtual filesystem, a single Python but install packages into a separate prefix/folder in the file system. This would require some work, as in particular the `venv` stdlib module is not available, and unlike with virtualenv it's not possible to have multiple Python executables in this use case. Still, I imagine it should be possible and it would be lighter than option 1.
 3. Finally, it's possible to have a single set of packages, but different `globals` objects and different states, which would be even faster, but then you can't indeed have the multiple versions of the same library installed, so it would likely not address this issue.

**Comment by fpliger** [comment]
It sounds that different options could serve different use cases. Would be good to understand the *need* for the many use cases, the actual effort in each option, and rough benchmark (to understand how much we actually get with each one) to help prioritize

**Comment by nixjdm** [comment]
Yes, precisely. Better IPC would be great, but the DOM would be sufficient
for cross WASM communication.

**Comment by marimeireles** [comment]
Seems like this issue is outdated as we removed `py-env` and are now using `py-config`.
I had a short conversation with part of the dev team (Nicholas, Madhur, Ted) and they agreed this wasn't something was worth pursuing for now.
If there's still interest from the community part please re-open this issue as a discussion. :)
Closing for now.
